src/t_hash.c

```diff
@@ -297,25 +296,403 @@ static void hashDictWithExpireOnRelease(dict *d) {
     ebDestroy(&dictExpireMeta->hfe, &hashFieldExpireBucketsType, NULL);
 }
 
+/*-----------------------------------------------------------------------------
+ * listpackTTL functions
+ *----------------------------------------------------------------------------*/
+/*
+ * If 'hexpire' command is called on a listpack hash object for the first time,
+ * we convert it to OBJ_ENCODING_LISTPACK_TTL encoding. Each field-value pair
+ * will have an additional TTL entry. So, listpack will have triplets in it:
+ * field-value-ttl. If TTL is not set for a field, we store 'zero' as the TTL
+ * value. 'zero' is encoded as two bytes in the listpack. Memory overhead of a
+ * non-existing TTL will be two bytes per field.
+ *
+ * Fields in the listpack will be ordered by TTL. Field with the smallest expiry
+ * time will be the first item. Fields without TTL will be at the end of the
+ * listpack. This way, it is easier/faster to find expired items.
+ */
+
+#define HASH_LP_NO_TTL 0
+
+/* Data structure for OBJ_ENCODING_LISTPACK_TTL. It contains listpack and
+ * metadata fields for hash field expiration.*/
+typedef struct listpackTTL {
+    ExpireMeta meta;  /* To be used in order to register the hash in the
+                         global ebuckets (i.e. db->hexpires) with next,
+                         minimum, hash-field to expire. */
+    sds key;          /* reference to the key, same one that stored in
+                         db->dict. Will be used from active-expiration flow
+                         for notification and deletion of the object, if
+                         needed. */
+    void *lp;         /* listpack that contains 'key-value-ttl' tuples which
+                         are ordered by ttl. */
+} listpackTTL;
```

Comment:
it's better to add `b->hexpires` reference to it, otherwise when defragging the listpack, we can't know which db or db->expires the listpack is belong to.

Reply:
okay sure. How do you know it in case of dict?

Reply:
get it from the metadata of dict.

Reply:
sorry, i misunderstood, the hfe ebuckets of the dict is for dict fields, so listpackTTL doesn't need it.